compiler/rustc_codegen_ssa/src/mir/intrinsic.rs

```diff
@@ -524,8 +524,19 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         };
 
                         let ty = substs.type_at(0);
-                        if int_type_width_signed(ty, bx.tcx()).is_some() {
-                            bx.atomic_rmw(atom_op, args[0].immediate(), args[1].immediate(), order)
+                        if int_type_width_signed(ty, bx.tcx()).is_some()
+                            || (ty.is_unsafe_ptr() && op == "xchg")
```

Comment:
It feels a bit piecemeal to do this just for one operation. Won't we simply be back at this soon with another operation?

If you have the time, could you remove all `as *mut usize` and `as usize` from the implementation of `AtomicPtr` and adjust the intrinsics accordingly?

Reply:
I don't think there is anything left (though I might have missed something). Remaining operations are intentionally unsupported in atomic ptr, so I don't think we should support them in intrinsics either.

Reply:
ah, great, thanks for having a look!

Reply:
Looks like this was not [quite right](https://github.com/rust-lang/miri/issues/1574#issuecomment-1053752658) (or the situation changed since then)

Reply:
There are no pointer to integer cast in AtomicPtr implementation.

Reply:
Hm, you are right... need to follow up in that issue then.